specification/cognitiveservices/data-plane/Language/preview/2022-02-01-preview/textanalytics.json

```diff
@@ -0,0 +1,2851 @@
+  },
+  "definitions": {
+    "AnalyzeTextTaskKind": {
+      "type": "string",
+      "description": "Enumeration of supported Text Analysis tasks.",
+      "enum": [
+        "SentimentAnalysis",
+        "EntityRecognition",
+        "PiiEntityRecognition",
+        "KeyPhraseExtraction",
+        "LanguageDetection",
+        "EntityLinking"
+      ],
+      "x-ms-enum": {
+        "name": "AnalyzeTextTaskKind",
+        "modelAsString": true
+      }
+    },
+    "AnalyzeTextLROTaskKind": {
+      "type": "string",
+      "description": "Enumeration of supported long-running Text Analysis tasks.",
+      "enum": [
+        "SentimentAnalysis",
+        "EntityRecognition",
+        "PiiEntityRecognition",
+        "KeyPhraseExtraction",
+        "EntityLinking",
+        "Healthcare",
+        "ExtractiveSummarization",
+        "CustomEntityRecognition",
+        "CustomSingleClassification",
+        "CustomMultiClassification"
+      ],
+      "x-ms-enum": {
+        "name": "AnalyzeTextLROTaskKind",
+        "modelAsString": true
+      }
+    },
+    "AnalyzeTextTaskResultsKind": {
+      "type": "string",
+      "description": "Enumeration of supported Text Analysis task results.",
+      "enum": [
+        "SentimentAnalysisResults",
+        "EntityRecognitionResults",
+        "PiiEntityRecognitionResults",
+        "KeyPhraseExtractionResults",
+        "LanguageDetectionResults",
+        "EntityLinkingResults"
+      ],
+      "x-ms-enum": {
+        "name": "AnalyzeTextTaskResultsKind",
+        "modelAsString": true
+      }
+    },
+    "AnalyzeTextLROResultsKind": {
+      "type": "string",
+      "description": "Enumeration of supported Text Analysis long-running operation task results.",
+      "enum": [
+        "SentimentAnalysisLROResults",
+        "EntityRecognitionLROResults",
+        "PiiEntityRecognitionLROResults",
+        "KeyPhraseExtractionLROResults",
+        "EntityLinkingLROResults",
+        "HealthcareLROResults",
+        "ExtractiveSummarizationLROResults",
+        "CustomEntityRecognitionLROResults",
+        "CustomSingleClassificationLROResults",
+        "CustomMultiClassificationLROResults"
+      ],
+      "x-ms-enum": {
+        "name": "AnalyzeTextLROResultsKind",
+        "modelAsString": true
+      }
+    },
+    "MultiLanguageAnalysisInput": {
+      "properties": {
+        "documents": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/MultiLanguageInput"
+          }
+        }
+      }
+    },
+    "LanguageDetectionAnalysisInput": {
+      "properties": {
+        "documents": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/LanguageInput"
+          }
+        }
+      }
+    },
+    "AnalyzeTextTask": {
+      "discriminator": "kind",
+      "required": [
+        "kind"
+      ],
+      "properties": {
+        "kind": {
+          "$ref": "#/definitions/AnalyzeTextTaskKind"
+        }
+      }
+    },
```

Comment:
This definition does not describe the whole request body for `analyze-text`, is that intentional? I see definitions for various tasks but they're not referenced anywhere.

Reply:
The problem here is that the body of the request for `analyze-text` is polymorphic, so we have to rely on the presence of the discriminator field `kind` to determine what the type is.

I discussed this previously with @johanste and my understanding was that this is the correct way to implement this, but I also feel a little bit uneasy about not having explicit `oneOf` semantics to rely on.

Reply:
I see. I agree that the body is polymorphic and must have a discriminator field, but I wonder why it currently only have the discriminator field?

Reply:
The reason it only has a discriminator field is because otherwise we would have to override any other parameters that we add in the base object in the derived objects, which is apparently something to be avoided here based on the previous discussions.

Reply:
Interesting. My main concern is that our code generators will not be able to generate a method with a useful input type for this API. It is a less of a concern for the Tier 1 languages because they all have convenience layers where this problem could be fixed manually so I am rather concerned about other languages when they decide to create auto-generated library for text analytics. I would like to hear @johanste and @tg-msft thoughts on this.

Reply:
Yes, that's the correct way to implement this per https://swagger.io/specification/v2/#models-with-polymorphism-support.  If there are any other common properties it'd be great to lift them up into the base definition.

AutoRest knows how to handle this and will generate something decent for all platforms _(though I'm sure it'll look a little funky in languages like Go without actual polymorphism)_.  There's no versioning concern with polymorphic input models, although they are harder for customers to discover via intellisense for some language so great documentation is a must.

Here's a similar example from Search and you can see the generated AutoRest code that turns this into a proper object hierarchy.
- [Swagger base definition](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/search/data-plane/Azure.Search/preview/2021-04-30-Preview/searchservice.json#L5891) vs [Generated base class](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/search/Azure.Search.Documents/src/Generated/Models/SearchIndexerDataIdentity.cs#L11)
- [Swagger derived definition](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/search/data-plane/Azure.Search/preview/2021-04-30-Preview/searchservice.json#L5913) vs [Generated derived class](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/search/Azure.Search.Documents/src/Generated/Models/SearchIndexerDataUserAssignedIdentity.cs#L13)
- [Swagger property definition](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/search/data-plane/Azure.Search/preview/2021-04-30-Preview/searchservice.json#L6077) vs [Generated property definition](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/search/Azure.Search.Documents/src/Generated/Models/SearchIndexerDataSourceConnection.cs#L50)